### ISSUE TITLE
Debrief audio: match race multi-channel behavior + cross-mic speaker diarization with progressive voice learning

### ISSUE THREAD

**Issue by weaties** [weaties/helmlog#648]
> ## Status — as of 2026-04-22
>
> **Partially shipped in #649 / 40ebe4b (merged to main).** The bug-fix
> scope + the diarization/auto-learn enhancement is live. The large
> remaining piece is the **voice-profile lifecycle state machine** and
> the **decision-table policy layer** from the structured spec comment
> below (gating extraction/auto-match on consent state, per-segment
> confidence scores, conflict flagging, full privacy fail-safe EARS
> invariants P1–P8). Those stay open; this issue tracks them.
>
> **Delivered in this slice:**
>
> - Debrief now uses the same recorder/config as race; debrief
>   `audio_sessions` rows match race on `channels`, `channel_map`, and
>   sibling-group shape (spec C1–C4, C6)
> - USB device-set change between race-end and debrief-start logs a
>   WARNING, doesn't block debrief (spec C5)
> - Per-channel pyannote diarization on race **and** debrief, dispatched
>   to the remote worker (spec decisions ✓). Position-prefixed labels
>   (`R1:SPEAKER_00`, `R2:SPEAKER_01`, …) keep cross-sibling labels
>   globally unique.
> - `num_speakers` hint exposed via `AUDIO_SPEAKERS_PER_CHANNEL` (default
>   2) so pyannote doesn't over-split the 2-mics-per-receiver setup
> - Voice embedding extraction keeps raw speaker labels end-to-end, so
>   `_try_auto_match` + `maybe_build_voice_profile` can key on the same
>   label segments store — cross-session auto-match is now wired
> - Click-to-assign crew on both race and debrief transcripts (label-
>   wide). Assign fans out to every sibling's speaker_map so the merged
>   view updates atomically.
> - Per-segment speaker override (new `transcript_segments.override_user_id`
>   column, migration v77) — a misattributed utterance can point at a
>   different crew member without mutating the label or touching other
>   segments.
> - Session page UX: unified transcript renderer (same block layout,
>   Follow toggle, pencil icon, auto-scroll synced to playhead) on both
>   race and debrief; streaming-mode audio path for long sibling
>   sessions (configurable via `AUDIO_STREAM_THRESHOLD_MINUTES`);
>   collapsible `Race Audio ▼` / `Debrief Audio ▼` section headers;
>   `R1` / `R2` isolation buttons on both players.
>
> **Not yet shipped (deferred to a follow-up):**
>
> - State-machine gating of voice-profile lifecycle (NoConsent /
>   ConsentedEmpty / Training / Active / Revoked) with the guard
>   conditions and revocation cleanup from Part 2 of the spec comment
> - Per-segment speaker-label decision table R1–R14 (confidence tiers,
>   conflict flagging, manual-review queue, mapped-vs-unmapped channel
>   rules) from Part 3
> - Privacy fail-safe invariants P1–P8 from Part 4 (strict consent
>   gating on embedding extraction, on-delete embedding purge,
>   no-consent segment fallback, etc.) — some are partially covered by
>   the existing consent model in #443 but haven't been audited
>   against the spec.
>
> The original issue body and structured spec comment follow below —
> left intact for context and to track the remaining scope. Do not
> treat the unchecked checkboxes below as "not done" without cross-
> referencing this status banner; several are already done.

---

## Summary

Two related problems, one issue because they share the same fix path:

1. **Bug** — debrief audio doesn't behave like race audio for multi-channel capture. We should be writing the same multi-channel WAV (with the same channel→position map) for debrief as we do for race sessions.
2. **Enhancement** — run speaker diarization on **both** race and debrief transcripts so a person can be identified regardless of which channel/microphone they're on, and have stored voice profiles build up automatically across sessions to auto-match future speakers.

## Background

Today's behavior (see `src/helmlog/transcribe.py`, `src/helmlog/audio.py`, `src/helmlog/routes/races.py`):

- The same recorder instance (`app.state.recorder`) is used for both race start (`/api/races/start`) and debrief start (`/api/races/{id}/debrief/start`). If the recorder is `AudioRecorderGroup` (sibling-card mode, #509) or a multi-channel `AudioRecorder` (#462), debrief should already capture the same channel layout. **In practice the debrief WAV does not have the same multi-channel structure** — needs investigation (is `AUDIO_CAPTURE_MODE` getting flipped, is sibling detection skipping during debrief, is the channel map missing for the debrief audio_session row, etc.).
- `_transcribe_multi_channel()` (`transcribe.py:125`) **intentionally skips pyannote diarization** for multi-channel audio, on the assumption that "hardware isolation already ties each channel to a person → speaker = position_name". That assumption holds during a race (everyone on station, mic on) but **breaks down during debrief**: crew gather around, share mics, take mics off, talk over the wrong channel. So a debrief transcript labeled `helm: ...` may actually be the tactician speaking into the helm's discarded mic.
- Voice-profile auto-matching (`_try_auto_match`, `transcribe.py:557`) only runs on **single-channel** transcripts. Multi-channel transcripts never contribute to or benefit from the `crew_voice_profiles` table built in #443.

## Decisions

- **Multi-channel parity for debrief is unconditional** — no opt-in, no separate mic configuration. Debrief uses whatever the race used.
- **Per-channel diarization, not mixdown** — preserves the channel→position cue and helps suppress crosstalk (a voice bleeding through a neighbor's mic gets diarized away from the channel's expected wearer).
- **Run per-channel diarization in parallel, offloaded to the Mac via `TRANSCRIBE_URL`** — N pyannote runs on a Pi serially is unworkable; the existing remote-offload path is the right place to absorb the cost.

### Bug fix — debrief multi-channel parity
- [x] Verify whether debrief sessions actually capture all configured channels/sibling cards. If not, fix `api_start_debrief` (or `capture_start`) so the same recorder configuration that runs for race start also runs for debrief start.
- [x] Persist the debrief `audio_sessions` row with the same `channels`, `channel_map`, and (where applicable) `capture_group_id`/`capture_ordinal` shape as race rows, so downstream multi-channel transcription and per-channel playback (#462) work identically.
- [x] Add an integration / smoke test that starts a race → ends → starts debrief and asserts the two audio_session rows have matching channel topology.

### Enhancement — diarization on multi-channel + cross-mic speaker matching
- [x] Run pyannote diarization **per channel** on multi-channel and sibling-card transcripts (don't trust the position label as ground-truth speaker).
- [x] Dispatch the per-channel diarization jobs in parallel to the Mac worker (`TRANSCRIBE_URL`); the Pi does not run pyannote locally for the multi-channel path.
- [x] When a channel's diarized speaker matches the channel's configured `position_name` (the expected wearer), tag the segment with that crew member directly — this is the high-confidence labeled training data we feed back into the voice profile.
- [x] When a channel's diarized speaker does **not** match the expected wearer, fall through to voice-profile auto-matching (`auto_match_speakers`) using the cross-session embedding store. This is the common case for debrief.
- [ ] Allow voice profiles to be built from multi-channel race audio: if channel `helm` is mapped to user `Alice` (via boat crew positions), every segment on that channel where diarization confirms a single dominant speaker is a labeled sample for Alice's profile. Today this requires manual transcript-level speaker assignment, which never happens because the multi-channel UI shows position names, not crew names. *(Partially done — click-to-assign is wired; automatic consent-gated voice profile accumulation from race audio is the remaining work.)*
- [ ] Surface and persist confidence scores on auto-matched speakers in both race and debrief views; allow manual correction (already wired for single-channel via #443) on multi-channel too. *(Manual correction done via per-segment override; persisted confidence scores in transcript_segments + UI badges not yet done.)*

### History / progressive learning
- [ ] Voice profiles should grow from race audio (clean, labeled training data) and apply to debrief audio (noisy, unlabeled test data). *(Infrastructure is in place — `_try_auto_match` + `maybe_build_voice_profile` work on the new prefixed labels. Needs end-to-end validation over a few sessions.)*
- [ ] Thresholds (`_AUTO_MATCH_HIGH = 0.7`, `_AUTO_MATCH_LOW = 0.4`) and minimums (`_MIN_SEGMENTS_FOR_PROFILE`, `_MIN_SESSIONS_FOR_PROFILE`) should be revisited once cross-channel data is feeding the profiles — likely tunable per-boat.
- [x] Existing `crew_voice_profiles` consent gating (#443, biometric PII per `docs/data-licensing.md`) must continue to apply: no embedding extraction or auto-matching for crew without `voice_profile` consent. *(Carried from #443; not newly strengthened — Part 4 of the spec adds defense-in-depth invariants P1–P8 that are still open.)*

## Risk tier

**High** — touches `transcribe.py` (transcription correctness), `audio.py` (capture topology), and `crew_voice_profiles` (biometric PII). Requires `/data-license` review for biometric handling and TDD per `tdd` skill. Spec recommended (`/spec`) given the combinatorial logic of (multi-channel × sibling × debrief × consent × confidence threshold) → speaker label.

## Related

- #462 — multi-channel hardware isolation (closed)
- #443 — diarized transcripts + voice learning (closed)
- #509 — sibling-card parallel capture (closed)
- #546 — debrief audio surfaced on session page (closed)
- #42 — original transcription + diarization roadmap (closed)
- #649 — this slice (merged 2026-04-22)

**Comment by weaties** [comment]
- Diarization on multi-channel adds significant CPU/GPU cost (today we skip it). Run it serially per channel, or in parallel? Is the existing `TRANSCRIBE_URL` remote offload path the right place to absorb the cost?

A: parallel - offload to mac

- Should we mix down to a single mono/stereo file specifically for diarization to avoid N pyannote runs? Single-pass diarization on a mixdown gives global speaker labels but loses the channel→position cue. Per-channel diarization preserves the cue but doesn't directly cluster the same voice across channels -- embeddings would need to be re-clustered globally before matching.

A: let's keep it per channel to better eliminate talking on top of each other.

- For the bug-fix scope: is there ever a case where debrief should intentionally use a different mic setup than race (e.g., a single shore-side mic at the dock)? If so, multi-channel parity needs to be opt-in/configurable, not unconditional.

A: not at the moment

**Comment by weaties** [comment]
## Structured Spec

**Risk tier:** High (transcribe.py, audio.py, crew_voice_profiles biometric PII)
**Spec format:** Mixed — EARS invariants + state diagram + decision table

---

### Part 1 — EARS: capture topology invariants (bug-fix scope)

| # | Requirement |
|---|---|
| C1 | THE SYSTEM SHALL use the same `recorder` and `audio_config` instance for debrief start as for race start within a given process. |
| C2 | WHEN a debrief is started after a race on the same device set, THE SYSTEM SHALL persist the debrief `audio_sessions` row with the same `channels` value as the prior race's `audio_sessions` row. |
| C3 | WHEN `AUDIO_CAPTURE_MODE=sibling`, THE SYSTEM SHALL create a new `capture_group_id` for the debrief with the same number of sibling ordinals (0..N-1) as the race. |
| C4 | WHEN a channel map exists for the session identity or admin default, THE SYSTEM SHALL attach the same `channel_map` to the debrief `audio_sessions` row as to the race row. |
| C5 | IF the USB device set changes between race end and debrief start (hot-unplug, new enumeration), THEN THE SYSTEM SHALL log a WARNING and record the debrief with whatever channel topology is currently available — debrief is not blocked. |
| C6 | THE SYSTEM SHALL expose a pytest-level integration test that starts a race → ends it → starts a debrief and asserts the two `audio_sessions` rows have identical `channels`, `channel_map`, and sibling-group shape. |

---

### Part 2 — State diagram: per-user voice-profile lifecycle

```mermaid
stateDiagram-v2
    [*] --> NoConsent : user exists, no consent
    NoConsent --> ConsentedEmpty : user grants voice_profile consent
    ConsentedEmpty --> Training : first labeled sample received
    Training --> Active : segments >= MIN_SEGMENTS AND sessions >= MIN_SESSIONS
    Training --> Training : labeled sample added (below threshold)
    Active --> Active : labeled sample added (centroid recomputed)
    ConsentedEmpty --> Revoked : consent revoked
    Training --> Revoked : consent revoked
    Active --> Revoked : consent revoked
    Revoked --> NoConsent : profile row + embeddings deleted
    NoConsent --> ConsentedEmpty : user re-grants consent
```

**Guard conditions:**
- `Training → Active`: requires both thresholds (`_MIN_SEGMENTS_FOR_PROFILE=30`, `_MIN_SESSIONS_FOR_PROFILE=2`) — tunable per-boat per the issue.
- `* → Revoked`: on revoke, existing embeddings and profile row MUST be deleted (not just marked), per `docs/data-licensing.md` biometric-consent revocation clause.
- `Revoked → NoConsent`: terminal cleanup state; must not retain any residual biometric data.
- Auto-matching (Part 3) runs ONLY while the user's profile state is `Active`.

---

### Part 3 — Decision table: per-segment speaker-label resolution

Applied to every whisper segment on every channel, after per-channel diarization. Abbreviations: `mapped_user` = the crew member that the channel's `position_name` resolves to via boat crew settings. `other_user` = any different crew member whose profile exists in `Active` state. **Embedding confidence**: `strong ≥ 0.7`, `marginal 0.4–<0.7`, `weak < 0.4`.

| Row | Channel mapped to crew? | Diarized: single dominant speaker on segment? | Embedding vs mapped_user | Embedding vs best other_user | mapped_user has `voice_profile` consent? | Applied label | Confidence | Feeds mapped_user profile? | Requires manual review? |
|---|---|---|---|---|---|---|---|---|---|
| R1 | yes | yes | strong | n/a | yes | `mapped_user.name` | high | **yes (race only)** | no |
| R2 | yes | yes | marginal | none/weaker | yes | `mapped_user.name` | medium | **yes (race only)** | no |
| R3 | yes | yes | marginal | strong (other, consented) | yes | `other_user.name` | medium | no | **yes** — conflict |
| R4 | yes | yes | weak | strong (other, consented) | yes | `other_user.name` | high | no | no |
| R5 | yes | yes | weak | none | yes | `position_name` (display) / `SPEAKER_XX` (machine) | low | no | no |
| R6 | yes | yes | no profile yet | n/a | yes | `mapped_user.name` (race) / `SPEAKER_XX` (debrief) | medium (race) / low (debrief) | **yes (race only)** | no |
| R7 | yes | no (overlap / multi-speaker) | n/a | n/a | yes | `SPEAKER_OVERLAP` | low | no | **yes** |
| R8 | yes | yes | any | strong (other, **no consent**) | yes | `position_name` | low | no | no |
| R9 | yes | yes | any | any | **no** | `position_name` (no crew linkage) | display-only | no | no |
| R10 | no (unmapped) | yes | n/a | strong (consented) | n/a | `other_user.name` | high | no | no |
| R11 | no (unmapped) | yes | n/a | marginal (consented) | n/a | `SPEAKER_XX` | low | no | **yes** |
| R12 | no (unmapped) | yes | n/a | none | n/a | `SPEAKER_XX` | low | no | no |
| R13 | no (unmapped) | no (overlap) | n/a | n/a | n/a | `SPEAKER_OVERLAP` | low | no | **yes** |
| R14 | n/a | n/a | n/a | n/a | n/a — single-channel audio (no position map) | existing `_try_auto_match` path (R10–R12 equivalent) | per auto-match | no | no |

**Key rules embedded:**
- "Feeds voice profile" is **race-only** (R1, R2, R6). Debrief segments are never used as training data — they are the noisy test set where the hardware-isolation assumption breaks down, per the issue decisions.
- A "conflict" case (R3) — channel says one person, voice says another, both plausible — is flagged for manual review rather than silently overriding.
- Consent gating is strict: R8 refuses to auto-apply a cross-mic match if the matched user has no consent (their profile shouldn't exist in Active state anyway, but defense-in-depth). R9 handles the expected-wearer-no-consent case by falling back to position label with no crew-user linkage.

---

### Part 4 — EARS: privacy and fail-safe invariants

| # | Requirement |
|---|---|
| P1 | THE SYSTEM SHALL refuse to extract a voice embedding for any user whose profile state is not `ConsentedEmpty`, `Training`, or `Active`. |
| P2 | WHEN a user revokes `voice_profile` consent, THE SYSTEM SHALL delete the `crew_voice_profiles` row and any cached embedding bytes within the request that processes the revocation, before returning success. |
| P3 | WHILE a user's profile state is `Revoked`, THE SYSTEM SHALL NOT use any residual embedding data for auto-matching, and MUST NOT retain embedding bytes on disk or in memory beyond the revocation transaction. |
| P4 | IF a segment's best embedding match is to a user without `voice_profile` consent, THEN THE SYSTEM SHALL NOT apply that user's name to the segment (R8). |
| P5 | IF `TRANSCRIBE_URL` is unreachable for a multi-channel session, THEN THE SYSTEM SHALL log a WARNING and fall back to the existing position-label-only behavior — no local pyannote run on the Pi for the multi-channel path, per the issue decisions. |
| P6 | THE SYSTEM SHALL dispatch per-channel diarization requests to the remote worker concurrently (not serially), bounded by a configurable parallelism limit (default = channel count). |
| P7 | THE SYSTEM SHALL record the applied confidence score on every non-display-only label (R1–R6, R10–R12) in `transcript_segments` so downstream UI can render confidence badges and the manual-review queue can filter by confidence. |
| P8 | WHEN a manual speaker correction is persisted by a crew member, THE SYSTEM SHALL NOT feed the corrected segment back into the training set unless the corrected user has consent AND the correction was made on race audio (not debrief). |

---

### Test-case generation

- **Part 1 EARS (C1–C6):** 6 direct tests, each asserting the named invariant. C6 is itself an integration test fixture.
- **Part 2 state diagram:** 1 test per transition (8 transitions + 2 re-entrant self-transitions) = 10 tests.
- **Part 3 decision table:** 14 tests (one per row), each a parameterised case with a fixture crew of 3 users (mapped, other, other-no-consent).
- **Part 4 EARS (P1–P8):** 8 direct tests, with P2/P3 requiring a disk-state assertion after revocation.

Total: ~38 targeted tests before any refactoring cleanup.

---

### Open implementation questions (not blockers for spec approval)

1. Should `SPEAKER_OVERLAP` be a distinct label or just "multiple `SPEAKER_XX`s on the same time range"? Affects UI rendering of overlap segments.
2. Where does the "conflict" manual-review flag live — a new column on `transcript_segments` or a separate `transcript_review_queue` table?
3. Is the parallelism cap from P6 a per-session setting or per-boat? Suggest per-boat with a sane default.

---

*Generated by `/spec`. Review and approve before implementation begins. Each row in a decision table = one test case. Each state transition = one test case.*

**Comment by weaties** [comment]
In progress on `feature/648-debrief-multichannel-parity` (Claude Code on MacBook-Pro-3.local) — scope: Part 1 (C1–C6) debrief multi-channel parity only. Enhancement parts 2–4 (voice-profile lifecycle, speaker-label decision table) to follow in a separate issue/PR.

**Comment by weaties** [comment]
Part 1 (C1–C6) + the follow-ups from field testing merged in #649 / 40ebe4b. Issue stays open for Parts 2–4 (voice-profile lifecycle, speaker-label decision table, privacy invariants). See the PR body for the delivered scope.
